FreqAI Strategy - Improve user experience

wagnercosta opened the pull request:
## Summary

The purpose of this PR is to improve the user experience when creating a freqAI strategy. Today the `populate_any_indicators` function can be confusing for a new user. This will also resolve a bug in the backtest that will allow it to more closely simulate the dry/run live behavior (see #7925)

Below are the main ideas:

- Split the functions of `feature_engineering` and `set_targets`
- Transport logic to the backend and avoid the user having to implement logic related to the parameters of a freqAI strategy (for example: `label_period_candles` and `include_shifted_candles`)

At first there will be 3 functions related to feature_engineering (the user only needs to implement 1 of them). The 3 are only necessary if the user wants more flexibility in defining the features).

**_These function names are tentative. I appreciate more intuitive name ideas._**

- [x] Code first draft version
- [x] Implement backtest with the new functions
- [x] Tests with classifier model
- [x] Docs
- [x] Tests

wagnercosta committed:
freqAI Strategy - improve user experience

wagnercosta commented:
Using the standard example strategy, below how the features would be allocated in each function:

## - feature_engineering_expand_all
```py
# extracted from `populate_any_indicators`
for t in self.freqai_info["feature_parameters"]["indicator_periods_candles"]:

            t = int(t)
            informative[f"%-{pair}rsi-period_{t}"] = ta.RSI(informative, timeperiod=t)
            informative[f"%-{pair}mfi-period_{t}"] = ta.MFI(informative, timeperiod=t)
            informative[f"%-{pair}adx-period_{t}"] = ta.ADX(informative, timeperiod=t)
            informative[f"%-{pair}sma-period_{t}"] = ta.SMA(informative, timeperiod=t)
            informative[f"%-{pair}ema-period_{t}"] = ta.EMA(informative, timeperiod=t)

            bollinger = qtpylib.bollinger_bands(
                qtpylib.typical_price(informative), window=t, stds=2.2
            )
            informative[f"{pair}bb_lowerband-period_{t}"] = bollinger["lower"]
            informative[f"{pair}bb_middleband-period_{t}"] = bollinger["mid"]
            informative[f"{pair}bb_upperband-period_{t}"] = bollinger["upper"]

            informative[f"%-{pair}bb_width-period_{t}"] = (
                informative[f"{pair}bb_upperband-period_{t}"]
                - informative[f"{pair}bb_lowerband-period_{t}"]
            ) / informative[f"{pair}bb_middleband-period_{t}"]
            informative[f"%-{pair}close-bb_lower-period_{t}"] = (
                informative["close"] / informative[f"{pair}bb_lowerband-period_{t}"]
            )

            informative[f"%-{pair}roc-period_{t}"] = ta.ROC(informative, timeperiod=t)

            informative[f"%-{pair}relative_volume-period_{t}"] = (
                informative["volume"] / informative["volume"].rolling(t).mean()
            )
```

```py
 def feature_engineering_expand_all(self, dataframe, period, **kwargs):
        dataframe["%-rsi-period"] = ta.RSI(dataframe, timeperiod=period)
        dataframe["%-mfi-period"] = ta.MFI(dataframe, timeperiod=period)
        dataframe["%-adx-period"] = ta.ADX(dataframe, timeperiod=period)
        dataframe["%-sma-period"] = ta.SMA(dataframe, timeperiod=period)
        dataframe["%-ema-period"] = ta.EMA(dataframe, timeperiod=period)

        bollinger = qtpylib.bollinger_bands(
            qtpylib.typical_price(dataframe), window=period, stds=2.2
        )
        dataframe["bb_lowerband-period"] = bollinger["lower"]
        dataframe["bb_middleband-period"] = bollinger["mid"]
        dataframe["bb_upperband-period"] = bollinger["upper"]

        dataframe["%-bb_width-period"] = (
            dataframe["bb_upperband-period"]
            - dataframe["bb_lowerband-period"]
        ) / dataframe["bb_middleband-period"]
        dataframe["%-close-bb_lower-period"] = (
            dataframe["close"] / dataframe["bb_lowerband-period"]
        )

        dataframe["%-roc-period"] = ta.ROC(dataframe, timeperiod=period)

        dataframe["%-relative_volume-period"] = (
            dataframe["volume"] / dataframe["volume"].rolling(period).mean()
        )
        
        return dataframe
```

## - feature_engineering_expand_basic

```py
# extracted from `populate_any_indicators`
        informative[f"%-{pair}pct-change"] = informative["close"].pct_change()
        informative[f"%-{pair}raw_volume"] = informative["volume"]
        informative[f"%-{pair}raw_price"] = informative["close"]
```

```py
    def feature_engineering_expand_basic(self, dataframe, **kwargs):
        dataframe["%-pct-change"] = dataframe["close"].pct_change()
        dataframe["%-raw_volume"] = dataframe["volume"]
        dataframe["%-raw_price"] = dataframe["close"]
        return dataframe
```

## - feature_engineering_standard
```py
# extracted from `populate_any_indicators`
            df["%-day_of_week"] = (df["date"].dt.dayofweek + 1) / 7
            df["%-hour_of_day"] = (df["date"].dt.hour + 1) / 25
```

```py
    def feature_engineering_standard(self, dataframe, **kwargs):     
        dataframe["%-day_of_week"] = (dataframe["date"].dt.dayofweek + 1) / 7
        dataframe["%-hour_of_day"] = (dataframe["date"].dt.hour + 1) / 25
        return dataframe
```

## - set_freqai_targets
```py
# extracted from `populate_any_indicators`
            df["&-s_close"] = (
                df["close"]
                .shift(-self.freqai_info["feature_parameters"]["label_period_candles"])
                .rolling(self.freqai_info["feature_parameters"]["label_period_candles"])
                .mean()
                / df["close"]
                - 1
            )
```

```py
    def set_freqai_targets(self, dataframe, **kwargs):
        dataframe["&-s_close"] = (
            dataframe["close"]
            .shift(-self.freqai_info["feature_parameters"]["label_period_candles"])
            .rolling(self.freqai_info["feature_parameters"]["label_period_candles"])
            .mean()
            / dataframe["close"]
            - 1
            )
        return dataframe
```

robcaulk committed:
improve doc, update test strats, change function names

robcaulk reviewed (changes requested):
It is looking good, but I do not see what happened to the `{pair}` in the column strings. Your docstrings indicate that these functions are automatically run for all `include_corr_pairs` but I do not see where you automatically add this to the feature strings. 

I made some changes to the naming convention, added documentation, updated documentation, and fixed the test strats. Please ensure my changes all align with your original goals. 

edit: looks like I missed the line that adds `{pair}_{tf}` to the suffix. Everything is looking good to me so far.

robcaulk committed:
ensure RL works with new naming scheme

wagnercosta commented:
> It is looking good, but I do not see what happened to the `{pair}` in the column strings. Your docstrings indicate that these functions are automatically run for all `include_corr_pairs` but I do not see where you automatically add this to the feature strings.
> 
> I made some changes to the naming convention, added documentation, updated documentation, and fixed the test strats. Please ensure my changes all align with your original goals.
> 
> edit: looks like I missed the line that adds `{pair}_{tf}` to the suffix. Everything is looking good to me so far.

Thanks for all the improvements, documentation, and testing. Now I will adjust the backtest feature.

wagnercosta commented on freqtrade/freqai/RL/BaseReinforcementLearningModel.py:
```diff
@@ -280,22 +280,26 @@ def build_ohlc_price_dataframes(self, data_dictionary: dict,
-                     f'%-{pair}raw_high_{tf}', f'%-{pair}raw_close_{tf}']
-        rename_dict = {f'%-{pair}raw_open_{tf}': 'open', f'%-{pair}raw_low_{tf}': 'low',
-                       f'%-{pair}raw_high_{tf}': ' high', f'%-{pair}raw_close_{tf}': 'close'}
+        ohlc_list = [f'%-raw_open_gen_{pair}_{tf}', f'%-raw_low_gen_{pair}_{tf}',
```
@robcaulk This will be a breaking change for RL strategies, right?

robcaulk replied:
Good point - we can add a check for the old and new strings. I will try to add and test that now :-)

wagnercosta committed:
adjust backtest to new feature engineering functions

wagnercosta committed:
freqai bt - fix tests

robcaulk commented on freqtrade/freqai/data_kitchen.py:
```diff
@@ -1188,7 +1287,91 @@ def use_strategy_to_populate_indicators(
+            for tf in tfs:
+                if tf not in base_dataframes:
+                    base_dataframes[tf] = pd.DataFrame()
+                if not corr_dataframes.keys():
```
Running live/dry with the "`new_version`" is throwing a `KeyError` for the `corr_dataframes` dict. I think it is due to this line. 

Here is the error:

```
  File "/home/rcaulk/projects/freqtrade/user_data/strategies/ReinforcementLearningExample5ac.py", line 134, in populate_indicators
    dataframe = self.freqai.start(dataframe, metadata, self)
  File "/home/rcaulk/projects/freqtrade/freqtrade/freqai/freqai_interface.py", line 142, in start
    dk = self.start_live(dataframe, metadata, strategy, self.dk)
  File "/home/rcaulk/projects/freqtrade/freqtrade/freqai/freqai_interface.py", line 421, in start_live
    dataframe = dk.use_strategy_to_populate_indicators(
  File "/home/rcaulk/projects/freqtrade/freqtrade/freqai/data_kitchen.py", line 1335, in use_strategy_to_populate_indicators
    dataframe = self.populate_features(dataframe.copy(), corr_pair, strategy,
  File "/home/rcaulk/projects/freqtrade/freqtrade/freqai/data_kitchen.py", line 1250, in populate_features
    informative_df = self.get_pair_data_for_features(
  File "/home/rcaulk/projects/freqtrade/freqtrade/freqai/data_kitchen.py", line 1201, in get_pair_data_for_features
    dataframe = corr_dataframes[pair][tf]
KeyError: '15m'
```

It is very odd, when I debug into this exact line from `freqai_interface.py`, I confirm that upon entry into this function, the `corr_dataframes` dict is not set to the default value (despite `freqai_interface.py line 421` not passing any key values for `corr_dataframes`). Instead, `corr_dataframes` does have the corr_pairlist keys, but only a single tf (`3m` in this case). However it is missing the higher `include_timeframes`. Since you have this line here, it is preventing the function from adding those keys. Which throws the error posted above. 

It seems you added this line for some fix in the bt tests? I guess I will let you decide how to ameliorate this since your head is deeper into the new backtesting updates.

wagnercosta replied:
Fixed! Thanks

robcaulk committed:
move price assignment to feature_engineering_standard() to reduce un-requested feature additions in RL. Ensure old method of price assignment still works, add deprecation warning to help users migrate their strategies

wagnercosta committed:
remove add pair to column from docs, fix keyerror bug and adjust hybrid strategy example

wagnercosta commented on freqtrade/freqai/freqai_interface.py:
```diff
@@ -301,18 +301,41 @@ def start_backtesting(
-                self.dd.load_metadata(dk)
-                dk.find_features(dataframe)
-                self.check_if_feature_list_matches_strategy(dk)
+                # self.dd.load_metadata(dk)
```
@robcaulk Can we remove this features check? I modified the logic and we don't have the features anymore if it's using backtesting saved predictions.

wagnercosta commented on freqtrade/freqai/freqai_interface.py:
```diff
@@ -301,18 +301,41 @@ def start_backtesting(
-                self.check_if_feature_list_matches_strategy(dk)
+                # self.dd.load_metadata(dk)
+                # dk.find_features(dataframe)
+                # self.check_if_feature_list_matches_strategy(dk)
```
@robcaulk  Can we remove this features check? I modified the logic and we don't have the features anymore if it's using backtesting saved predictions.

robcaulk replied:
I am not sure, we still want to make sure the strategy is providing the same features that were used to create the saved backtesting predictions. If the strategy has changed the features, we want them to be aware that the use of the saved backtesting predictions is no longer reflecting the current strat.

Does this make sense? or am I missing something important?

wagnercosta replied:
Yes. I'll think in a best way to check this without impact the performance.

wagnercosta replied:
Done.

wagnercosta committed:
added again feature check in BT from pred files

robcaulk committed:
Update freqai-feature-engineering.md

paranoidandy committed:
Update FreqaiExampleStrategy.py

Change can_short to True to enable shorting

wagnercosta committed:
fix get dataframe data to include startup_candle

wagnercosta committed:
fix get dataframe data to include startup_candle

wagnercosta committed:
Merge branch 'develop' into freqai_feature_engineering_functions

wagnercosta committed:
fix corr_pairs startup candle count bug

wagnercosta commented on docs/freqai-running.md:
```diff
@@ -67,6 +67,12 @@ Backtesting mode requires [downloading the necessary data](#downloading-data-to-
     *want* to retrain a new model with the same config file, you should simply change the `identifier`.
     This way, you can return to using any model you wish by simply specifying the `identifier`.
 
+!!! Note
```
@robcaulk Can you help improving this note, please?

robcaulk committed:
improve wording in freqai doc

wagnercosta committed:
fix tests

wagnercosta committed:
Merge branch 'develop' into freqai_feature_engineering_functions

xmatthias committed:
Merge pull request #7971 from paranoidandy/patch-1

Update FreqaiExampleStrategy.py

robcaulk committed:
ensure cached corr-pairs works with new framework

robcaulk reviewed (dismissed):
This is ready to go, thanks for the hard work @wagnercosta .

From a coverage perspective - I will just remind that it is decreased because we are now carrying around a bunch of additional code to support users through the transition. This code should be removed in a couple months. By then - all the older users will have seen the deprecation warning and will have had an opportunity to transition.

robcaulk committed:
improve deprecation note

robcaulk reviewed (dismissed):
LGTM

xmatthias reviewed (changes requested):
If i'm honest, then i think we'll need better "migration documentation" (similar to migration.md), properly explaining "this part from the existing strategy goes there, this part goes here".

keep in mind, this is a breaking change (at least going forward - with an expiration date) - so we must provide users with a clear migration path.
Doesn't have to be anything extensive - a simple "this block in existing method" goes to "this new method" should suffice.
something around `migration.md` (either place there - or in a new file dedicated for freqAI) - really simple - highlight parts - and where they move to (in my understanding, it's not much other change) - or if there's other changes in the logic - outlining these.

The new docs are good for someone starting out / from scratch - but provide basically 0 guidance on how to migrate from an existing and working freqAI strategy (though we show "you must migrate until ..." in the logs).

> From a coverage perspective 

I'm not too worried about the 0.1% drop in coverage - freqAI is not VERY well covered as is (82% in develop) - but existing tests already take quite a long time (at least a few) - so i'm fine with not having full coverage of freqAI code, as more coverage would almost certainly mean more (potentially even slower) tests.

xmatthias commented on freqtrade/freqai/data_kitchen.py:
```diff
@@ -1188,7 +1287,94 @@ def use_strategy_to_populate_indicators(
+                           "Please update your strategy by using "
+                           "the new feature_engineering functions. See \n"
+                           "https://www.freqtrade.io/en/latest/freqai-feature-engineering/"
+                           "for details.")
```
we MUST throttle this - either to "on startup" (Where `new_version` is detected) - or to "once per base timeframe".

otherwise it'll be way too verbose - dropping this message for every call - obscuring all other logs (assuming a user would like

wagnercosta replied:
done

xmatthias committed:
Add FreqAI migration documentation

xmatthias commented:
@robcaulk @wagnercosta can you have a look at the last commit please (make sure my logic / explanations make sense)?

it's adding the migration docs as mentioned above. Please make sure to look at it in rendered mode (`mkdocs serve` - go to localhost:8000).
otherwise you might miss how it's actually looking:

![image](https://user-images.githubusercontent.com/5024695/211476947-a11d4630-2c20-4c0a-89d8-9d3bbe6dd0f9.png)


Unfortunately, we can't highlight in different colors - so we're using annotations to show where which part goes.

wagnercosta committed:
moved deprecated warning to start function

wagnercosta commented:
> @robcaulk @wagnercosta can you have a look at the last commit please (make sure my logic / explanations make sense)?

@xmatthias LGTM. Thanks for handle this.

robcaulk commented on docs/strategy_migration.md:
````diff
@@ -477,3 +477,254 @@ after:
+        return df
+```
+
+1. Features - Move to `feature_engineering_expand_all`
````
Wow that is a cool feature. Thanks a lot for that!

robcaulk reviewed (approved):
LGTM

xmatthias reviewed (approved):
LGTM :+1:
